lib/pwnlib/reg_sort.rb

```diff
@@ -0,0 +1,221 @@
+# encoding: ASCII-8BIT
+
+require 'pwnlib/context'
+
+module Pwnlib
+  # Topographical sort
+  module RegSort
+    # @note Do not create and call instance method here. Instead, call module method on {RegSort}.
+    module ClassMethod
+      # Walk down the assignment list of a register,
+      # return the path walked if it is encountered again.
+      #
+      # @return [Array]
+      #   The registers that involved in the cycle start from `reg`.
+      #   Empty array will be returned if no cycle start and end at `reg`.
+      # @example
+      #   check_cycle('a', {a: 1}) #=> []
+      #   check_cycle('a', {a: 'a'}) #=> ['a']
+      #   check_cycle('a', {a: 'b', b: 'c', c: 'b', d: 'a'}) #=> []
+      #   check_cycle('a', {a: 'b', b: 'c', c: 'd', d: 'a'})
+      #   #=> ['a', 'b', 'c', 'd']
+      def check_cycle(reg, assignments)
```

Comment:
Also, we probably should do input hash normalization (keys to_s) in regsort directly, and just assume that Hash is `{String => String}` in all helper functions and simplify all these to_s things...  (Would need to fix example for this function too)